Fixing tests for graph convolutions, DAG models with TF 2.2RC

rbharath opened the pull request:
Working to debug convolutions...

rbharath commented:
Unfortunately, tests didn't just start passing with TF2.2 RC, so I'll have to go in and do a more thorough debugging pass. Will report on here as that progresses...

rbharath commented:
I've fixed a few of the failures. There was an issue with the `KerasModel` class having `model.inputs` as None, and the atomic convs were failing since a default parameter on `tf.gather` changed.

The main issue remaining now is that `GraphConvModel` is broken. Interesting, the `GraphConv` layer itself seems to be functional with tests passing. I'm not yet sure what about the actual model is broken but taking a look.

peastman commented on deepchem/models/keras_model.py:
```diff
@@ -173,7 +173,7 @@ def _create_inputs(self, example_inputs):
     self._ensure_built()
     self._inputs_built = True
-    if len(self.model.inputs) > 0:
+    if (self.model.inputs is not None) and len(self.model.inputs) > 0:
```
Is this a change in TF 2.2?  In 2.1 and earlier, if a model doesn't define inputs then its `inputs` field is an empty list rather than None.  If so, search for all uses of `model.inputs` to make sure all of them will handle it correctly.

rbharath replied:
I think it is a change. The `GraphConvKerasModel` class has `model.inputs` set to `None` upon construction. I went through the file and searched for all uses to make sure it was handled correctly.

peastman commented on deepchem/models/keras_model.py:
```diff
@@ -295,7 +295,7 @@ def fit_generator(self,
     var_key = None
     if variables is not None:
-      var_key = tuple(v.experimental_ref() for v in variables)
+      var_key = tuple(v.ref() for v in variables)
```
Note this will only work in TF 2.2 and later.

rbharath commented:
I'm seeing the following error from the graph conv models:

```
WARNING:tensorflow:Model was constructed with shape (None,) for input Tensor("input_4:0", shape=(None,), dtype=int32), but it was called on an input with incompatible shape ().
WARNING:tensorflow:Model was constructed with shape (None,) for input Tensor("input_5:0", shape=(None,), dtype=float32), but it was called on an input with incompatible shape ().
Traceback (most recent call last):
  File "graph_conv_script.py", line 133, in <module>
    model(inputs)
  File "/Users/bharath/opt/anaconda3/envs/deepchem/lib/python3.6/site-packages/tensorflow/python/keras/engine/base_layer.py", line 967, in __call__
    outputs = self.call(cast_inputs, *args, **kwargs)
  File "/Users/bharath/opt/anaconda3/envs/deepchem/lib/python3.6/site-packages/tensorflow/python/keras/engine/network.py", line 714, in call
    convert_kwargs_to_constants=base_layer_utils.call_context().saving)
  File "/Users/bharath/opt/anaconda3/envs/deepchem/lib/python3.6/site-packages/tensorflow/python/keras/engine/network.py", line 894, in _run_internal_graph
    assert str(id(x)) in tensor_dict, 'Could not compute output ' + str(x)
AssertionError: Could not compute output Tensor("softmax/Identity:0", shape=(None, 1, 2), dtype=float32)
```
I think it's something about how the inputs are handled that's changed subtly. I'm able to run the model directly in eager mode (without using `tf.keras.layers.Input` and feeding in arrays directly). I'm hitting a wall now though so will pick this back up tomorrow

rbharath commented:
I've got a simple minimal failing test case now. The following script fails:

```
import numpy as np
import deepchem as dc
import tensorflow as tf
from deepchem.feat.mol_graphs import ConvMol
from tensorflow.keras.layers import Input 
from deepchem.models import layers
from deepchem.molnet import load_bace_classification 

featurizer='GraphConv'
_, all_dataset, _ = load_bace_classification(featurizer)
train, _, _ = all_dataset
dataset = dc.data.NumpyDataset(train.X[:10])

(X_b, _, _, _) = next(dataset.iterbatches(
    batch_size=50,
    deterministic=True,
    pad_batches=True))
multiConvMol = ConvMol.agglomerate_mols(X_b)
inputs = [
    multiConvMol.get_atom_features(), multiConvMol.deg_slice,
    np.array(multiConvMol.membership)
]
inputs += multiConvMol.get_deg_adjacency_lists()[1:]

number_atom_features=75
atom_features = Input(shape=(number_atom_features,))
degree_slice = Input(shape=(2,), dtype=tf.int32)
membership = Input(shape=tuple(), dtype=tf.int32)

deg_adjs = []
for i in range(0, 10 + 1):
  deg_adj = Input(shape=(i + 1,), dtype=tf.int32)
  deg_adjs.append(deg_adj)

gc1_in = [atom_features, degree_slice, membership] + deg_adjs
gc1 = layers.GraphConv(64, activation_fn=tf.nn.relu)(gc1_in)

model = tf.keras.Model(
    inputs=[
        atom_features, degree_slice, membership
    ] + deg_adjs,
    outputs=[gc1])

model(inputs)
```
However, this next script runs just fine:
```
import numpy as np
import deepchem as dc
import tensorflow as tf
from deepchem.feat.mol_graphs import ConvMol
from tensorflow.keras.layers import Input, Dense, Reshape
from deepchem.models import layers
from deepchem.molnet import load_bace_classification 

featurizer='GraphConv'
_, all_dataset, _ = load_bace_classification(featurizer)
train, _, _ = all_dataset
dataset = dc.data.NumpyDataset(train.X[:10])

(X_b, _, _, _) = next(dataset.iterbatches(
    batch_size=50,
    deterministic=True,
    pad_batches=True))
multiConvMol = ConvMol.agglomerate_mols(X_b)
inputs = [
    multiConvMol.get_atom_features(), multiConvMol.deg_slice,
    np.array(multiConvMol.membership)
]
inputs += multiConvMol.get_deg_adjacency_lists()[1:]


gc1 = layers.GraphConv(64, activation_fn=tf.nn.relu)(inputs)
```

Somehow the `GraphConv` layer doesn't play well with being put into `tf.keras.Model`. @peastman do you have any ideas what might be going on?

rbharath commented:
Forgot to attach the error message. The first script returns:

```
(deepchem) bharath@Bharaths-MBP tests % python simple_graph_conv_scrip.py
Loading dataset from disk.
Loading dataset from disk.
Loading dataset from disk.
2020-03-24 14:03:19.339062: I tensorflow/core/platform/cpu_feature_guard.cc:143] Your CPU supports instructions that this TensorFlow binary was not compiled to use: AVX2 FMA
2020-03-24 14:03:19.351283: I tensorflow/compiler/xla/service/service.cc:168] XLA service 0x7faeb9d6b7e0 initialized for platform Host (this does not guarantee that XLA will be used). Devices:
2020-03-24 14:03:19.351303: I tensorflow/compiler/xla/service/service.cc:176]   StreamExecutor device (0): Host, Default Version
Traceback (most recent call last):
  File "simple_graph_conv_scrip.py", line 43, in <module>
    model(inputs)
  File "/Users/bharath/opt/anaconda3/envs/deepchem/lib/python3.6/site-packages/tensorflow/python/keras/engine/base_layer.py", line 967, in __call__
    outputs = self.call(cast_inputs, *args, **kwargs)
  File "/Users/bharath/opt/anaconda3/envs/deepchem/lib/python3.6/site-packages/tensorflow/python/keras/engine/network.py", line 714, in call
    convert_kwargs_to_constants=base_layer_utils.call_context().saving)
  File "/Users/bharath/opt/anaconda3/envs/deepchem/lib/python3.6/site-packages/tensorflow/python/keras/engine/network.py", line 894, in _run_internal_graph
    assert str(id(x)) in tensor_dict, 'Could not compute output ' + str(x)
AssertionError: Could not compute output Tensor("graph_conv/Identity:0", shape=(None, 64), dtype=float32)
```

rbharath commented:
I think what's happening is that `tf.keras.Model` is getting confused because this model doesn't follow clean batching. Here's the shape of the input tensors on a single batch:

```
inputs[0].shape                                                                                                                   
(1984, 75)                                                                                                                        
inputs[1].shape                                                                                                                   
(11, 2)                                                                                                                           
inputs[2].shape                                                                                                                   
(1984,)                                                                                                                           
inputs[3].shape                                                                                                                   
(405, 1)                                                                                                                          
inputs[4].shape                                                                                                                   
(943, 2)
inputs[5].shape
(581, 3)
inputs[6].shape
(55, 4)
inputs[7].shape
(0, 5)
inputs[8].shape
(0, 6)
inputs[9].shape
(0, 7)
inputs[10].shape
(0, 8)
inputs[11].shape
(0, 9)
inputs[12].shape
(0, 10)
```
I think things are of different sizes which confuses the input layers.

rbharath commented:
I think I have a partial fix. The basic idea is that the graph-conv model needs to be written with TensorFlow's subclassing API instead of the usual functional/symbolic API. This makes it possible to train/predict models, but it makes other things break, like the neural fingerprints. I think model saving is also different.

I'll keep working on this, but it's going to take some more time before this is ready...

rbharath commented:
Down to one failing test now. I've fixed the graph convs, but the DAG model is still broken. Starting work on debugging it, but probably another day or two's effort needed

rbharath commented:
I've got the DAG models running again, but they're not learning anything, so one of the overfit tests is failing. Hitting a wall for now, but will keep poking at it

rbharath commented:
Ok, tests now appear to be passing, so I think this PR is ready for review. Here's a summary of the changes made here:

- `GraphConvModel` uses non-standard control flow. It takes inputs of different sizes for each batch, some of which can be empty. This was confusing TF2.X's functional mode (which assumes everything comes cleanly in batches). I converted this class to use the imperative subclassing API which allows for more flexible control flow. I introduce a new class `GraphConvKerasModel` which directly inherits `tf.keras.Model` to do this.
- The change to imperative style means that neural fingerprints retrieval can no longer work the way it used to. The `predict(..,outputs=[...])` style only works for functional mode and not imperative mode. The neural fingerprint is a type of embedding, and it seems likely that we'd want models to be able to emit embeddings, so rather than make a special hack for neural fingerprint extract, I added a new top level function `KerasModel.predict_embedding(...)`. If you label an `output_type` with 'embedding`, you can retrieve it by calling `predict_embedding`. This may be useful for other models where we want to extract embeddings.
- `DAGModel` previously used `tf.compat.v1.scatter_nd_update` which updated Variables in places. To take advantage of this feature, it made a new variable upon each invocation of the function. This no longer works in TF2.X, since tf.function calls can't make new variables after their first invocation. I used the new `tf.tensor_scatter_nd_update` which does the same update on a tensor to get rid of the variable and just use a tensor. `DAGModel` learns very slowly though, so I had to boost the number of epochs to get it to overfit correctly.
- `DAGLayer`, `DAGGather`, `DAGTransformer` didn't have tests. In the process of debugging, I ended up writing unit tests to help me that are probably worth having.
- This PR currently runs with TF 2.2rc. I haven't tested whether these changes would run under TF2.1. We would either need to wait until the proper TF2.2 release to release DeepChem 2.4.0, or figure out how to make the tests pass with 2.1. I'm personally inclined to just wait till TF2.2 releases since we'll have to make that upgrade anyways.

We should review these changes thoroughly since there's a lot of things in this PR. @peastman, @VIGS25 would both of you mind taking a close look at this PR and letting me know your feedback?

peastman commented on deepchem/models/graph_models.py:
```diff
@@ -495,6 +504,115 @@ def default_generator(self,
         ], [y_b], [w_b])
 
 
+class GraphConvKerasModel(tf.keras.Model):
```
If this is meant to be an internal class, should there be a `_` at the start of the name?

rbharath replied:
Good suggestion. Will put an `_` at the start.

peastman commented on deepchem/models/tests/test_overfit.py:
```diff
@@ -645,7 +646,7 @@ def test_DAG_singletask_regression_overfit(self):
 
     # Fit trained model
-    model.fit(dataset, nb_epoch=50)
+    model.fit(dataset, nb_epoch=1200)
```
If it worked before with 50 epochs, and now it needs 1200, it's a pretty good bet something is broken!

rbharath replied:
I checked the original paper, and they say they trained for 5000 epochs to converge on the Delaney dataset. I did a few passes over the code, and I'm not seeing anything that's jumping out as wrong after a few hours of scrutiny. It's definitely possible something is broken here, but I'm hitting a wall on this one.

peastman commented:
> I added a new top level function KerasModel.predict_embedding(...). If you label an output_type with 'embedding, you can retrieve it by calling predict_embedding`.

I'm wondering if this is overly specific.  Embeddings are just one example of a case where someone might want to retrieve an internal layer.  For example, what if a model can include arbitrary values in its `output_types` field, and then you can pass strings as the `outputs` parameter when you call `predict()` to tell it to return the outputs of a particular type?  GraphConvModel could still have a `predict_embedding()` method, which would just be a convenience function for `predict(..., outputs='embedding')`.

rbharath commented:
I like your suggestion of including arbitrary values in `output_types`. This could be broadly useful for making models with non-standard transformations. I'll change the implementation along the lines you suggest!

rbharath commented:
Something's weird here; Travis CI isn't running tests. I'm not sure if this is a Travis bug, but will poke at it some more.

rbharath commented:
As I'm testing this branch more deeply, I'm seeing a number more issues. Many models looks to converge more slowly than they used to on TF1.X. I'm doing a more thorough testing pass to feel confident that models actually work which will take some more time, so marking this back as [WIP]

peastman commented:
I've been seeing the same problem on the OpenMM repo.  The travis tests actually do get run, but the results don't appear on the github pull request.  You can see the tests for recent PRs at https://travis-ci.org/github/deepchem/deepchem/pull_requests.  I haven't tried to track down the problem yet.

rbharath commented:
Something strange that's going on here: Rebasing on top of https://github.com/deepchem/deepchem/pull/1787 breaks a bunch of the tests. Many models train much more slowly, with the graph convs that took 10 epochs to overfit now taking 50. Weave models that ran correctly now produce NaNs after the rebase. 

As far as I can tell, that PR introduces an innocuous bugfix to set training flags for dropout, but it seems to break a whole bunch of things. I've verified that removing the flag lets the tests run correctly once again.

@peastman Do you have any ideas what might be happening?

peastman commented:
Are they overfit tests that are supposed to run without dropout?  If so, make sure they're really disabling it.  Setting the training flag may also be affecting the behavior of the BatchNormalization layers.

rbharath commented:
I've fixed the issues that popped up after the rebase by making sure that Dropout/Batchnorm were handled correctly in training. I also went through and added a couple more tests and more docstrings in various places.

I've also changed the implementation of `predict_embedding` as suggested above. It's possible that `DAGModel` still has issues, but I'm hitting a wall there. I suggest we document the issue, then go ahead and merge this one in

CC @peastman @vsomnath

rbharath commented:
I'm going to go ahead and merge this in. I'll open a new issue to track remaining TF2.X conversion issues.
